mockgcp/mocklogging/logmetric.go

```diff
@@ -0,0 +1,179 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mocklogging
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/timestamppb"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+
+	"github.com/GoogleCloudPlatform/k8s-config-connector/mockgcp/common/projects"
+	pb "github.com/GoogleCloudPlatform/k8s-config-connector/mockgcp/generated/mockgcp/logging/v2"
+	"github.com/golang/protobuf/ptypes/empty"
+)
+
+type metricsService struct {
+	*MockService
+	pb.UnimplementedMetricsServiceV2Server
+}
+
+func (s *metricsService) GetLogMetric(ctx context.Context, req *pb.GetLogMetricRequest) (*pb.LogMetric, error) {
+	name, err := s.parseLogMetricName(req.MetricName)
+	if err != nil {
+		return nil, err
+	}
+
+	fqn := name.String()
+
+	obj := &pb.LogMetric{}
+	if err := s.storage.Get(ctx, fqn, obj); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, status.Errorf(codes.NotFound, "metric %q not found", name)
+		} else {
+			return nil, status.Errorf(codes.Internal, "error reading metric: %v", err)
+		}
+	}
+
+	return obj, nil
+}
+
+func (s *metricsService) CreateLogMetric(ctx context.Context, req *pb.CreateLogMetricRequest) (*pb.LogMetric, error) {
+	reqName := req.Parent + "/metrics/" + req.GetMetric().GetName()
+	name, err := s.parseLogMetricName(reqName)
+	if err != nil {
+		return nil, err
+	}
+
+	fqn := name.String()
+
+	now := time.Now()
+
+	obj := proto.Clone(req.GetMetric()).(*pb.LogMetric)
+	obj.Name = name.MetricName
+	obj.CreateTime = timestamppb.New(now)
+	obj.UpdateTime = timestamppb.New(now)
+
+	// Not clear why, but we return the values before populating the MetricDescriptor Name/Type
+	retval := proto.Clone(obj).(*pb.LogMetric)
+
+	s.populateDefaultsForLogMetric(name, obj)
+
+	if err := s.storage.Create(ctx, fqn, obj); err != nil {
+		return nil, status.Errorf(codes.Internal, "error creating metric: %v", err)
```

Comment:
Do we want to treat already exists errors similar to NotFound but return an AlreadyExists error?

Reply:
Yeah, maybe what I'll do is look into OCC and stop using apierrors and start using status.Errorf with codes (so we can just `return nil, err` (and check the code first if we need to)

Reply:
I might do that in a separate PR though, as it will break all the open PRs and the ones that have yet to be opened.

Reply:
Fair enough, it just seems odd that we care about the distinction of resource existence errors from internal errors on some of these APIs but not others.